### DISCUSSION TITLE
Is the IdleTimeout setting description correct?

### DISCUSSION THREAD

**Discussion by bentoi** [microsoft/msquic#3875] [General]
The documentation of [the `IdleTimeout` setting](https://github.com/microsoft/msquic/blob/main/docs/Settings.md) explains that the connection is closed gracefully if the connection is idle after `IdleTimeout`. Is this really correct? 

The Quic's spec specifies that the connection should be silently closed if it's idle. The msquic implementation also appears to perform a silent close https://github.com/microsoft/msquic/blob/4bcd6e89f1e26c93f3b1bf8e8285675d6ee5f668/src/core/connection.c#L6151 when the idle timer is triggered.

As written today, it sounds like the `IdleTimeout` is about gracefully closing inactive connections but afaict it's about detecting a connection loss when the connection has outstanding packets waiting to be ACKed. Is this correct?

**Comment by nibanks** [comment]
The documentation is wrong. It should be "silently shut down" instead of "gracefully shut down". Thanks for the find!

**Reply by bentoi** [reply]
Ok thanks. One other related question... How is the application supposed to set the `KeepAliveInterval` if it doesn't know the negotiated idle timeout? I would think it should be set to a value inferior to the negotiated idle timeout.

**Reply by nibanks** [reply]
Keep alives are generally meant for keeping the path alive, not getting around idle timeouts. Idle timeouts are essentially saying "I don't want to keep this connection open, without work, for longer than X." Keep alives are saying "I don't want the network (i.e., NAT) to lose this connection". So, you usually set keep alives to something like 20 or 30 seconds. Idle timeouts are completely up your app layer to decide, if and how long.

**Reply by bernardnormier** [reply]
Ok, this makes sense. 

But then, how do you see an http/3 implementation that uses msquic perform the following:

> HTTP clients are expected to request that the transport keep connections open while there are responses outstanding for requests or server pushes, as described in [Section 10.1.2](https://www.rfc-editor.org/rfc/rfc9000#section-10.1.2) of [[QUIC-TRANSPORT](https://datatracker.ietf.org/doc/html/rfc9000)].

That's from https://datatracker.ietf.org/doc/html/rfc9114#name-idle-connections

Does msquic provide an API other than `KeepAliveInterval` that allow the application to send QUIC PING frames?

See also: https://github.com/dotnet/runtime/issues/87478

We are particularly interested in this topic because we implement an RPC over QUIC (currently .NET QUIC - see https://docs.icerpc.dev/). We don't want the idle timer to close a connection with a single outstanding request waiting for the corresponding response; a request/response maps to a bidir QUIC stream with icerpc.

**Reply by nibanks** [reply]
HTTP/3 clients would dynamically enable keep alives while they have an outstanding request, so that the server can respond without fear of the path being killed (i.e., every ~20 seconds). When the request completes (success/failure/timeout) keep alives are then disabled.

The connection isn't considered really idle because there is work, but the client is just waiting for it. So the connection will stay alive as long as it takes for the server to respond.

**Reply by bernardnormier** [reply]
> HTTP/3 clients would dynamically enable keep alives

They would do so by setting `KeepAliveInterval` on the connection, or does msquic provide another API?

>  so that the server can respond without fear of the path being killed (i.e., every ~20 seconds)

The concern here is the negotiated idle timeout, not a well-known timeout value like 20 seconds. The http/3 client needs to send keep alives (PING frames) at intervals lower than this negotiated idle timeout.

Is it possible to retrieve this negotiated idle timeout?

**Reply by nibanks** [reply]
> They would do so by setting KeepAliveInterval on the connection, or does msquic provide another API?

Yes.

> The concern here is the negotiated idle timeout, not a well-known timeout value like 20 seconds. The http/3 client needs to send keep alives (PING frames) at intervals lower than this negotiated idle timeout.
> 
> Is it possible to retrieve this negotiated idle timeout?

Not currently, but that'd be super easy to expose via some new `GetParam` for the connection, something like `QUIC_PARAM_CONN_NEGOTIATED_IDLE_TIMEOUT`. If you'd like this, please open a feature request. That being said, if a peer does this, (1) they are explicitly not wanting to keep a connection open long, and (2) it's most likely simply a configuration bug, and you might not want to cover that up by setting super aggressive keep alives.

What do you want to do if the peer configured a 5 second idle timeout? Do you really want to configure 2-3 second keep alives?

**Reply by bernardnormier** [reply]
> What do you want to do if the peer configured a 5 second idle timeout? Do you really want to configure 2-3 second keep alives?

This should be up to the application. msquic could support a min value for the negotiated timeout (the connection establishment fails if the peer sends a lower max_idle_timeout).

Or the application could retrieve this negotiated idle timeout and abort the connection if it's too low.

> Not currently, but that'd be super easy to expose via some new GetParam for the connection, something like QUIC_PARAM_CONN_NEGOTIATED_IDLE_TIMEOUT. If you'd like this, please open a feature request.

My preference would be to have a  simpler API, where the application (= the application protocol using msquic) can instruct msquic to keep the connection alive. A simple on/off switch.

When "keep alive" is on, msquic figures out when / how often to send PING frames to the peer, based on the negotiated idle timeout. When "keep alive" is off, msquic doesn't send PING frames.

Does this sound reasonable?

**Reply by nibanks** [reply]
Please open a feature request with all the details. Thanks!

**Reply by bernardnormier** [reply]
See #3880
